metric/bigtable/src/main/java/com/spotify/heroic/metric/bigtable/api/RowRangeReader.java

```diff
@@ -0,0 +1,237 @@
+/*
+ * Copyright (c) 2016 Spotify AB.
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.heroic.metric.bigtable.api;
+
+import com.google.bigtable.v2.BigtableGrpc.Bigtable;
+import com.google.bigtable.v2.ColumnRange;
+import com.google.bigtable.v2.ReadRowsRequest;
+import com.google.bigtable.v2.ReadRowsResponse;
+import com.google.bigtable.v2.ReadRowsResponse.CellChunk;
+import com.google.bigtable.v2.ReadRowsResponse.CellChunk.RowStatusCase;
+import com.google.bigtable.v2.RowFilter;
+import com.google.bigtable.v2.RowSet;
+import com.google.protobuf.ByteString;
+
+import com.spotify.heroic.bigtable.grpc.stub.StreamObserver;
+import com.spotify.heroic.metric.bigtable.api.BigtableDataClient.CellConsumer;
+
+import eu.toolchain.async.AsyncFramework;
+import eu.toolchain.async.AsyncFuture;
+import eu.toolchain.async.ResolvableFuture;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+class RowRangeReader {
+
+    private static final int MAX_READ_BATCH_SIZE = 500000;
+
+    static class CellChunkHandler {
+
+        private final Consumer<List<CellChunk>> commitCells;
+        private final int consumedCells;
+        private final List<ByteString> partialCellCarryOver;
+        private final List<CellChunk> nonCommittedCellsCarryOver;
+
+        CellChunkHandler(Consumer<List<CellChunk>> commitCells) {
+            this(commitCells, 0, Collections.emptyList(), Collections.emptyList());
+        }
+
+        private CellChunkHandler(
```

Comment:
This would be the `@RequiredArgsConstructor(access=AccessLevel.PRIVATE)` annotation with lombok.